RFCs/FS-1060-nullable-reference-types.md

```diff
@@ -0,0 +1,823 @@
+| Non-nullable warnings | Off by default | On by default | Yes |
+| Warnings from other assemblies | Off by default | On by default | Yes |
+
+These kinds of warnings should be individually tunable on a per-project basis.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+### Complexity
+
+Although it could be argued that this simplifies reference types by making the fact that they could be `null` explicit, it does give programmers another thing that must explicitly account for. Although it fits within the "spirit" of F# to make nullability of reference types explicit, it's arguable that F# programmers need to account for enough things already.
+
+This is also a very complicated feature, with lots of edge cases, that has significantly reduced utility if any of the following are true:
+
+* There is insufficient flow analysis such that F# programmers are asserting non-nullability all over the place just to avoid unnecessary warnings
+* Existing, working code is infected without the programmer explicitly opting into this new behavior
+* The rest of the .NET ecosystem simply does not adopt non-nullability, thus making adornments the new normal
+
+Although we cannot control the third point, ensuring that the implementation is complete and up to spec with respect to what C# also does (while accounting for F#-isms) is the only way to do our part in helping the ecosystem produce less nulls.
+
+Additionally, we now have the following possible things to account for with respect to accessing underlying data:
+
+* Nullable reference types
+* Option
+* ValueOption
+* Choice
+* Result
+
+These are a lot of ways to send data to various parts of the system, and there is a danger that people could use nullable reference types for more than just the boundary levels of their system.
+
+### Unsoundness
+
+This feature is inherently unsound. For starters, warnings don't prevent people from flowing `null` through their code like errors do. Compare this with the F# option type, which offers zero way to attempt to coerce `None` into a `Some x`.
+
+More subtly, there is no way to extract nullability information with reflection today without a change in the .NET runtime. This means that it can be impossible to know, at least by reflection, if the type of something is actually nullable or non-nullable. This means that reflection-based libraries (such as JSON.NET) may be able to define a contract that they cannot fulfill; i.e., make an API's return type a `'T` when it could still be `null`.
+
+Finally, the "handles null" attribute can be trivially abused, and it may be likely that the F# compiler could be "tricked" into thinking something will no longer be `null`.
+
+All of this means that this feature is entrusting acting in good faith on the greater .NET ecosystem, rather than preventing acting in bad faith with compile-time guarantees.
+
+## Alternatives
+[alternatives]: #alternatives
+
+The primary alternative is to simply not do this.
```

Comment:
I think there are lots of other alternatives we should list. e.g. foremost is this:

1. Design a fully sound (apart from runtime casts/reflection, akin to F# units of measure) and fully checked non-nullness system for F# (which gives errors rather than warnings), then interoperate with C# (giving warnings around the edges and for interop)

Given that F# is in good shape w.r.t. nullness this is actually not such a silly thing to consider.

Another is

2. Allow code to be generic w.r.t. nullness (i.e. so you can express "if you give me a null in, I might give you a null back; if you give me a non-null in, I will give you a non-null back")

Reply:
Added